internal/olm/operator/registry/fbcindex/configMapWriter.go

```diff
@@ -0,0 +1,225 @@
+	ConfigMapEncodingAnnotationGzip = "gzip+base64"
+)
+
+/*
+This file implements the actual building of the CM list. It uses the template method design pattern to implement both
+regular string VM, and compressed binary CM.
+
+The method itself is FBCRegistryPod.getConfigMaps. This file contains the actual implementation of the writing actions,
+used by the method.
+*/
+
+type configMapWriter interface {
+	reset()
+	newConfigMap(string) *corev1.ConfigMap
+	getFilePath() string
+	isEmpty() bool
+	exceedMaxLength(cmSize int, data string) (bool, error)
+	closeCM(cm *corev1.ConfigMap) error
+	addData(data string) error
+	continueAddData(data string) error
+	writeLastFragment(cm *corev1.ConfigMap) error
+}
+
+type defaultCMWriter struct {
+	cmName      string
+	namespace   string
+	dataBuilder strings.Builder
+}
+
+func newDefaultWriter(name, namespace string) *defaultCMWriter {
+	return &defaultCMWriter{
+		cmName:      name,
+		namespace:   namespace,
+		dataBuilder: strings.Builder{},
+	}
+}
+
+func (cmw *defaultCMWriter) reset() {
+	cmw.dataBuilder = strings.Builder{}
+}
+
+func (cmw *defaultCMWriter) newConfigMap(name string) *corev1.ConfigMap {
+	return &corev1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: corev1.SchemeGroupVersion.String(),
+			Kind:       "ConfigMap",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: cmw.namespace,
+		},
+		Data: map[string]string{},
+	}
+}
+
+func (cmw *defaultCMWriter) isEmpty() bool {
+	return cmw.dataBuilder.Len() > 0
+}
+
+func (cmw *defaultCMWriter) exceedMaxLength(cmSize int, data string) (bool, error) {
+	return cmSize+cmw.dataBuilder.Len()+len(data) > maxConfigMapSize, nil
+}
+
+func (cmw *defaultCMWriter) closeCM(cm *corev1.ConfigMap) error {
+	cm.Data[defaultConfigMapKey] = cmw.dataBuilder.String()
+	cmw.dataBuilder = strings.Builder{}
+	return nil
+}
+
+func (cmw *defaultCMWriter) getFilePath() string {
+	return fmt.Sprintf("%s.yaml", defaultConfigMapKey)
+}
+
+func (cmw *defaultCMWriter) addData(data string) error {
+	cmw.dataBuilder.WriteString(data)
+	return nil
+}
+
+// continueAddData completes adding the data after starting adding it in exceedMaxLength
+func (cmw *defaultCMWriter) continueAddData(data string) error {
+	return cmw.addData(data)
+}
+
+func (cmw *defaultCMWriter) writeLastFragment(cm *corev1.ConfigMap) error {
+	cm.Data[defaultConfigMapKey] = cmw.dataBuilder.String()
+
+	return nil
+}
+
+type gzipCMWriter struct {
+	actualBuff   *bytes.Buffer
+	helperBuff   *bytes.Buffer
+	actualWriter *gzip.Writer
+	helperWriter *gzip.Writer
+	cmName       string
+	namespace    string
+}
+
+func newGZIPWriter(name, namespace string) *gzipCMWriter {
+	actualBuff := &bytes.Buffer{}
+	helperBuff := &bytes.Buffer{}
+
+	return &gzipCMWriter{
+		actualBuff:   actualBuff,
+		helperBuff:   helperBuff,
+		actualWriter: gzip.NewWriter(actualBuff),
+		helperWriter: gzip.NewWriter(helperBuff),
+		cmName:       name,
+		namespace:    namespace,
+	}
+}
```

Comment:
I still think that having multiple config map writer implementations is unnecessary. I think we could get way with doing something like:
```suggestion
type configMapWriter struct {
	actualBuff   *bytes.Buffer
	helperBuff   *bytes.Buffer
	actualWriter io.Writer
	helperWriter io.Writer
	cmName       string
	namespace    string
	buffLimit       int
}

func newConfigMapWriter(name, namespace string, gzipped bool) *configMapWriter {
	actualBuff := &bytes.Buffer{}
	helperBuff := &bytes.Buffer{}

	cmWriter := &configMapWriter{
		actualBuff:   actualBuff,
		helperBuff:   helperBuff,
		actualWriter: bufio.NewWriter(actualBuff)
		helperWriter: bufio.NewWriter(helperBuff)
		cmName:       name,
		namespace:    namespace,
		buffLimit: maxConfigMapSize,
	}
	
	if gzipped {
	    cmWriter.actualWriter = gzip.NewWriter(actualBuff)
	    cmWriter.helperWriter = gzip.NewWriter(helperBuff)
	    buffLimit = maxGZIPLength
	}
	
	return cmWriter
}
```

This lets us use the exact same implementation for both while just adjusting the writer and buffer limit used when the gzip flag is passed.

Reply:
I still think it's a very good implementation. I think the suggested solution above misses many of the differences between the two writers. The design pattern in use here is well known and it's a solid solution for similar challenges. 

There are several advantage for this approach. First, it detaches the partitioning algorithm from the actual writing, kipping most of the flow common and relatively easy to read. The actual write to the configMap, and all the rest of the specifications of actual chosen configMap type, are implemented elsewhere. 

The main flow does not have branches every other line, and each writer can be very tight and accurate.

Bottom line - as I see it - this approach is easier to implement, read and maintain, and follows well known design pattern.

Another advantage, that I don't think will be related here, is if we'll need to add a third type of configMap (e.g. json based), we'll only need to implement a new writer.

Reply:
> I still think it's a very good implementation. I think the suggested solution above misses many of the differences between the two writers. The design pattern in use here is well known and it's a solid solution for similar challenges.

Is there really a major difference between the default and gzip ConfigMap writer? From what I can tell it would just be the difference of using a `gzip.NewWriter` versus a `bufio.NewWriter` and all the underlying functions implement the same logic.

> There are several advantage for this approach. First, it detaches the partitioning algorithm from the actual writing, kipping most of the flow common and relatively easy to read. The actual write to the configMap, and all the rest of the specifications of actual chosen configMap type, are implemented elsewhere.
>
> The main flow does not have branches every other line, and each writer can be very tight and accurate.

I don't think the solution I suggested changes this though. It would still have a the same functional breakdown, but we would only be changing what implementation of the `io.Writer` interface we are using when we write the specified data.


> Bottom line - as I see it - this approach is easier to implement, read and maintain, and follows a well known design pattern.

I'm not saying that this is a bad or unreadable design pattern, I'm just questioning whether or not we actually need to have 2 separate implementations that implement the `configMapWriter` interface when it seems to me that the difference between the two becomes using a `gzip.NewWriter()` function vs a `bufio.NewWriter()` function to initialize the writer type. I think all the other `gzipConfigMapWriter` functions are general enough that we just swap out the `.actualWriter` and `.helperWriter` values to be whatever `io.Writer` implementation we want.

> Another advantage, that I don't think will be related here, is if we'll need to add a third type of configMap (e.g. json based configMap), we'll only need to implement a new writer.

We could still keep the interface around and have this ability (and having the interface makes it easier to test).

--- 

Overall I don't disagree with your points or the design of the implementation. My questioning is strictly around whether or not we *need* 2 separate implementations of the `configMapWriter` interface.

Reply:
So looking at this more it does look like there is the difference of *where* the data is written to in the ConfigMap (`ConfigMap.Data` vs `ConfigMap.BinaryData`) but I still think that could be generalized by an internal function used to write the data to the right ConfigMap field.

Reply:
Could we _always_ write to the binary data field?

Reply:
> Could we _always_ write to the binary data field?

I don't think so. textual data should be in the `Data` field.

Reply:
Thanks again @everettraven for you patience. I'm not sure I get your point. What wrong with multiple implementations for an interface? this is the whole idea. We **want** to have multiple implementations. I think we gain many advantages going this path. As I said, this is not my invention - this programming challenge is not new, nor its solutions. 

Specifically, changing the implementation of the default writer to work with two internal writer just to not having multiple implementation of the interface, is not a good reason, I think, because we will write every segment of the data twice. We don't need this for the default implementation, since we can know the length without writing the next segment. This is not the case for the compression implementation, where we must write the next segment to check the new length.

So the solution will work, but it will be less efficient.